Source/astcenccli_toplevel.cpp

```diff
@@ -399,6 +400,165 @@ static void test_inappropriate_cpu_extensions()
+ *
+ * @return The astc image file, or nullptr is the load failed.
+ */
+static astc_codec_image* load_decomp_file(
+	const char* filename,
+	unsigned int dim_z,
+	int padding,
+	bool y_flip,
+	bool linearize_srgb,
+	bool& img_is_hdr,
+	int& img_num_components
+) {
+	int load_result = 0;
+	astc_codec_image *image = nullptr;
+
+	// For a 2D image just load the image directly
+	if (dim_z == 1)
+	{
+		image = astc_codec_load_image(filename, padding, y_flip, linearize_srgb, &load_result);
+	}
+	else
+	{
+		int slice_result = 0;
+		astc_codec_image* slice = nullptr;
+		std::vector<astc_codec_image*> slices;
+
+		// For a 3D image load an array of slices
+		for (unsigned int image_index = 0; image_index < dim_z; image_index++)
+		{
+			bool error;
+			std::string slice_name = get_slice_filename(filename, image_index, error);
+			if (error)
+			{
+				printf("ERROR: Image pattern does not contain an extension: %s\n", filename);
+				break;
+			}
+
+			slice = astc_codec_load_image(slice_name.c_str(), padding, y_flip, linearize_srgb, &slice_result);
+			if (!slice)
+			{
+				break;
+			}
+
+			assert(slice_result != 0);
+			slices.push_back(slice);
+
+			// Check it is not a 3D image
+			if (slice->zsize != 1)
+			{
+				printf("ERROR: Image arrays do not support 3D sources: %s\n", slice_name.c_str());
+				break;
+			}
+
+			// Check slices are consistent with each other
+			if (image_index != 0)
+			{
+				if (load_result != slice_result)
+				{
+					printf("ERROR: Image array[0] and [%d] are different formats\n", image_index);
+					break;
+				}
+
+				if ((slices[0]->xsize != slice->xsize) ||
+				    (slices[0]->ysize != slice->ysize) ||
+				    (slices[0]->zsize != slice->zsize))
+				{
+					printf("ERROR: Image array[0] and [%d] are different dimensions\n", image_index);
+					break;
+				}
+			}
+			else
+			{
+				load_result = slice_result;
+			}
+		}
+
+		// If all slices loaded correctly then repack them into a single image
+		if (slices.size() == dim_z)
+		{
+			unsigned int dim_x = slices[0]->xsize;
+			unsigned int dim_y = slices[0]->ysize;
+			int bitness = (load_result & 0x80) ? 16 : 8;
+			int slice_size = (dim_x + (2 * padding)) * (dim_y + (2 * padding));
+
+			image = alloc_image(bitness, dim_x, dim_y, dim_y, padding);
+
+			// Combine 2D source images into one 3D image; skipping padding slices
+			for (unsigned int z = padding; z < dim_z + padding; z++)
+			{
+				if (bitness == 8)
+				{
+					size_t copy_size = slice_size * 4 * sizeof(uint8_t);
+					memcpy(*image->data8[z], *slices[z - padding]->data8[0], copy_size);
+				}
+				else
+				{
+					size_t copy_size = slice_size * 4 * sizeof(uint16_t);
+					memcpy(*image->data16[z], *slices[z - padding]->data16[0], copy_size);
+				}
+			}
+
+			// Fill in the padding slices with clamped data
+			fill_image_padding_area(image);
+		}
+
+		for (auto &i : slices)
+		{
+			free_image(i);
+		}
+	}
+
+	img_num_components = load_result & 7;
```

Comment:
Why are we multiplying it with 7? would be a good idea to have it as a const with a comment

Reply:
The original code returns a packed int (bit[2..0] = num components, bit[3]  = 16 bit if set). Seems like needless optimization, so made that long-hand in the next commit.